README.md
# DscLcmConfiguration

This gem helps build configuration scripts for the Windows PowerShell Desired State Configuration Local Configuration Manager.

It currently supports the initial release of WMF 4, WMF 4 with the unversioned update, and WMF 5.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'dsc_lcm_configuration'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install dsc_lcm_configuration

## Usage

```ruby
require 'dsc_lcm_configuration'

# This will create an object that can generate a WMF 4 (without the update)
# compatible LCM configuration.
DscLcmConfiguration::Factory.create()
```

`#create` takes two parameters, the first is the version of the LCM.

* 'wmf4' will create a WMF4 RTM compatible configuration
* 'wmf4_with_update' supports the additional LCM settings introduced in the unversioned update to WMF 4
* 'wmf5' supports WMF 5.0 RTM (not the preview builds)

The second parameter is an array of options.  Which options are supported depends on the version of WMF you are targeting.


## Development

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/smurawski/dsc_lcm_configuration.



CHANGELOG.md
# Change Log

## [v0.2.0](https://github.com/smurawski/dsc_lcm_configuration/tree/v0.2.0) (2016-06-23)
[Full Changelog](https://github.com/smurawski/dsc_lcm_configuration/compare/v0.1.0...v0.2.0)

**Merged pull requests:**

- Only Update Existing Config Settings [\#2](https://github.com/smurawski/dsc_lcm_configuration/pull/2) ([smurawski](https://github.com/smurawski))

## [v0.1.0](https://github.com/smurawski/dsc_lcm_configuration/tree/v0.1.0) (2016-06-23)
**Merged pull requests:**

- First release [\#1](https://github.com/smurawski/dsc_lcm_configuration/pull/1) ([smurawski](https://github.com/smurawski))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path("../../lib", __FILE__)
require "dsc_lcm_configuration"


spec/dsc_lcm_configuration_spec.rb
require "spec_helper"

describe DscLcmConfiguration do
  it "has a version number" do
    expect(DscLcmConfiguration::VERSION).not_to be nil
  end

  it "defaults to the LCM base config" do
    expect(DscLcmConfiguration::LcmBase).to receive(:new)
    expect(DscLcmConfiguration::LcmV4).to_not receive(:new)
    expect(DscLcmConfiguration::LcmV5).to_not receive(:new)
    DscLcmConfiguration::Factory.create()
  end

  it "returns LcmV4 for `wmf4_with_update`" do
    expect(DscLcmConfiguration::LcmBase).to_not receive(:new)
    expect(DscLcmConfiguration::LcmV4).to receive(:new)
    expect(DscLcmConfiguration::LcmV5).to_not receive(:new)
    DscLcmConfiguration::Factory.create("wmf4_with_update")
  end

  it "returns LcmV4 for `4`" do
    expect(DscLcmConfiguration::LcmBase).to_not receive(:new)
    expect(DscLcmConfiguration::LcmV4).to receive(:new)
    expect(DscLcmConfiguration::LcmV5).to_not receive(:new)
    DscLcmConfiguration::Factory.create("4")
  end

  it "returns LcmV5 for `wmf5`" do
    expect(DscLcmConfiguration::LcmBase).to_not receive(:new)
    expect(DscLcmConfiguration::LcmV4).to_not receive(:new)
    expect(DscLcmConfiguration::LcmV5).to receive(:new)
    DscLcmConfiguration::Factory.create("wmf5")
  end

  it "returns LcmV4 for `5`" do
    expect(DscLcmConfiguration::LcmBase).to_not receive(:new)
    expect(DscLcmConfiguration::LcmV4).to_not receive(:new)
    expect(DscLcmConfiguration::LcmV5).to receive(:new)
    DscLcmConfiguration::Factory.create("5")
  end
end


spec/dsc_lcm_configuration/lcm_v4_spec.rb
# -*- encoding: utf-8 -*-

require "spec_helper"

describe DscLcmConfiguration::LcmV4 do

  let(:base_config_settings) do
    {
      :action_after_reboot => "StopConfiguration",
      :allow_module_overwrite => false,
      :certificate_id => "$null",
      :configuration_mode => "ApplyAndAutoCorrect",
      :configuration_mode_frequency_mins => 30,
      :debug_mode => "All",
      :reboot_if_needed => false,
      :refresh_mode => "PUSH",
      :refresh_frequency_mins => 15,
    }
  end

  describe "Default LCM Properties" do

    let(:lcm) { DscLcmConfiguration::Factory.create("wmf4_with_update") }

    it "action_after_reboot defaults to StopConfiguration" do
      expect(lcm.action_after_reboot).to eq(base_config_settings[:action_after_reboot])
    end

    it "allow_module_overwrite defaults to false" do
      expect(lcm.allow_module_overwrite).to eq(base_config_settings[:allow_module_overwrite])
    end

    it "certificate_id defaults to $null" do
      expect(lcm.certificate_id).to eq(base_config_settings[:certificate_id])
    end

    it "configuration_mode defaults to ApplyAndAutoCorrect" do
      expect(lcm.configuration_mode).to eq(base_config_settings[:configuration_mode])
    end

    it "configuration_mode_frequency_mins defaults to 30" do
      expect(lcm.configuration_mode_frequency_mins).to eq(base_config_settings[:configuration_mode_frequency_mins])
    end

    it "debug_mode defaults to All" do
      expect(lcm.debug_mode).to eq(base_config_settings[:debug_mode])
    end

    it "reboot_if_needed defaults to false" do
      expect(lcm.reboot_if_needed).to eq(base_config_settings[:reboot_if_needed])
    end

    it "refresh_mode defaults to PUSH" do
      expect(lcm.refresh_mode).to eq(base_config_settings[:refresh_mode])
    end

    it "refresh_frequency_mins defaults to 15" do
      expect(lcm.refresh_frequency_mins).to eq(base_config_settings[:refresh_frequency_mins])
    end
  end

  describe "Base LCM Script" do
    let(:lcm) { DscLcmConfiguration::Factory.create("wmf4_with_update") }
    let(:base_config_script) do
      <<-EOH
        configuration SetupLCM
        {
          LocalConfigurationManager
          {
            ActionAfterReboot = 'StopConfiguration'
            AllowModuleOverwrite = [bool]::Parse('false')
            CertificateID = $null
            ConfigurationMode = 'ApplyAndAutoCorrect'
            ConfigurationModeFrequencyMins = 30
            DebugMode = 'All'
            RebootNodeIfNeeded = [bool]::Parse('false')
            RefreshFrequencyMins = 15
            RefreshMode = 'PUSH'
          }
        }
      EOH
    end

    it "generates a default LCM configuration script" do
      expect(lcm.lcm_configuration_script).to eq(base_config_script)
    end

    it "returns standard config options" do
      expect(lcm.lcm_config).to eq(base_config_settings)
    end
  end

  describe "With config" do
    let(:lcm_config) { { :allow_module_overwrite => true, :configuration_mode => "ApplyOnly" } }

    let(:lcm_with_config) { DscLcmConfiguration::Factory.create("wmf4_with_update", lcm_config) }
    let(:updated_config_script) do
      <<-EOH
        configuration SetupLCM
        {
          LocalConfigurationManager
          {
            ActionAfterReboot = 'StopConfiguration'
            AllowModuleOverwrite = [bool]::Parse('true')
            CertificateID = $null
            ConfigurationMode = 'ApplyOnly'
            ConfigurationModeFrequencyMins = 30
            DebugMode = 'All'
            RebootNodeIfNeeded = [bool]::Parse('false')
            RefreshFrequencyMins = 15
            RefreshMode = 'PUSH'
          }
        }
      EOH
    end
    let(:updated_config_settings) do
      {
        :action_after_reboot => "StopConfiguration",
        :allow_module_overwrite => true,
        :certificate_id => "$null",
        :configuration_mode => "ApplyOnly",
        :configuration_mode_frequency_mins => 30,
        :debug_mode => "All",
        :reboot_if_needed => false,
        :refresh_mode => "PUSH",
        :refresh_frequency_mins => 15,
      }
    end

    it "generates an updated LCM configuration script" do
      expect(lcm_with_config.lcm_configuration_script).to eq(updated_config_script)
    end

    it "returns updated config options" do
      expect(lcm_with_config.lcm_config).to eq(updated_config_settings)
    end
  end
end


lib/dsc_lcm_configuration/lcm_v4.rb
# -*- encoding: utf-8 -*-
#
# Author:: Steven Murawski (<steven.murawski@gmail.com>)
#
# Copyright (C) 2014 Steven Murawski
#
# Licensed under the Apache 2 License.
# See LICENSE for more details

require "dsc_lcm_configuration/lcm_base"

module DscLcmConfiguration
  class LcmV4 < LcmBase

    def lcm_properties
      {
        :action_after_reboot => "StopConfiguration",
        :allow_module_overwrite => false,
        :certificate_id => nil,
        :configuration_mode => "ApplyAndAutoCorrect",
        :configuration_mode_frequency_mins => 30,
        :debug_mode => "All",
        :reboot_if_needed => false,
        :refresh_mode => "PUSH",
        :refresh_frequency_mins => 15,
      }
    end

    def lcm_configuration_script
      <<-LCMSETUP
        configuration SetupLCM
        {
          LocalConfigurationManager
          {
            ActionAfterReboot = '#{action_after_reboot}'
            AllowModuleOverwrite = [bool]::Parse('#{allow_module_overwrite}')
            CertificateID = #{certificate_id}
            ConfigurationMode = '#{configuration_mode}'
            ConfigurationModeFrequencyMins = #{configuration_mode_frequency_mins}
            DebugMode = '#{debug_mode}'
            RebootNodeIfNeeded = [bool]::Parse('#{reboot_if_needed}')
            RefreshFrequencyMins = #{refresh_frequency_mins}
            RefreshMode = '#{refresh_mode}'
          }
        }
      LCMSETUP
    end
  end
end


lib/dsc_lcm_configuration/lcm_v5.rb
# -*- encoding: utf-8 -*-
#
# Author:: Steven Murawski (<steven.murawski@gmail.com>)
#
# Copyright (C) 2014 Steven Murawski
#
# Licensed under the Apache 2 License.
# See LICENSE for more details

require "dsc_lcm_configuration/lcm_base"

module DscLcmConfiguration
  class LcmV5 < LcmBase

    def lcm_properties
      {
        :action_after_reboot => "StopConfiguration",
        :allow_module_overwrite => false,
        :certificate_id => nil,
        :configuration_mode => "ApplyAndAutoCorrect",
        :configuration_mode_frequency_mins => 15,
        :debug_mode => "All",
        :reboot_if_needed => false,
        :refresh_mode => "PUSH",
        :refresh_frequency_mins => 30,
      }
    end

    def lcm_configuration_script
      <<-LCMSETUP
        [DSCLocalConfigurationManager()]
        configuration SetupLCM
        {
          Settings
          {
            ActionAfterReboot = '#{action_after_reboot}'
            AllowModuleOverwrite = [bool]::Parse('#{allow_module_overwrite}')
            CertificateID = #{certificate_id}
            ConfigurationMode = '#{configuration_mode}'
            ConfigurationModeFrequencyMins = #{configuration_mode_frequency_mins}
            DebugMode = '#{debug_mode}'
            RebootNodeIfNeeded = [bool]::Parse('#{reboot_if_needed}')
            RefreshFrequencyMins = #{refresh_frequency_mins}
            RefreshMode = '#{refresh_mode}'
          }
        }
      LCMSETUP
    end
  end
end


lib/dsc_lcm_configuration/lcm_base.rb
# -*- encoding: utf-8 -*-
#
# Author:: Steven Murawski (<steven.murawski@gmail.com>)
#
# Copyright (C) 2014 Steven Murawski
#
# Licensed under the Apache 2 License.
# See LICENSE for more details

module DscLcmConfiguration
  class LcmBase

    def self.lcm_factory(version = "wmf4", options)
      case version.to_s.downcase
      when "4", "wmf4_with_update"
        LcmV4.new(options)
      when "5", "wmf5"
        LcmV5.new(options)
      else
        LcmBase.new(options)
      end
    end

    def lcm_properties
      @lcm_properties ||= {
        :allow_module_overwrite => false,
        :certificate_id => nil,
        :configuration_mode => "ApplyAndAutoCorrect",
        :configuration_mode_frequency_mins => 30,
        :reboot_if_needed => false,
        :refresh_mode => "PUSH",
        :refresh_frequency_mins => 15,
      }
    end

    def initialize(config = {})
      @certificate_id = nil
      lcm_properties.each do |setting, value|
        send(setting, value)
      end

      config.each do |setting, value|
        send(setting, value) if lcm_properties.keys.include? setting
      end
    end

    def method_missing(name, *args)
      return super unless lcm_properties.keys.include?(name)
      if args.length == 1
        instance_variable_set("@#{name}", args.first)
      else
        instance_variable_get("@#{name}")
      end
    end

    def certificate_id(value = nil)
      if value.nil?
        @certificate_id.nil? ? "$null" : "'#{@certificate_id}'"
      else
        @certificate_id = value
      end
    end

    def lcm_config
      hash = {}
      lcm_properties.keys.each do |key|
        hash[key] = send(key)
      end
      hash
    end

    def lcm_configuration_script
      <<-LCMSETUP
        configuration SetupLCM
        {
          LocalConfigurationManager
          {
            AllowModuleOverwrite = [bool]::Parse('#{allow_module_overwrite}')
            CertificateID = #{certificate_id}
            ConfigurationMode = '#{configuration_mode}'
            ConfigurationModeFrequencyMins = #{configuration_mode_frequency_mins}
            RebootNodeIfNeeded = [bool]::Parse('#{reboot_if_needed}')
            RefreshFrequencyMins = #{refresh_frequency_mins}
            RefreshMode = '#{refresh_mode}'
          }
        }
      LCMSETUP
    end
  end
end
